gpu/src/cuda/shaders/u256.cuh

```diff
@@ -0,0 +1,152 @@
+// https://github.com/andrewmilson/ministark/blob/main/gpu-poly/src/metal/u256.h.metal
+
+#ifndef u256_h
+#define u256_h
+
+#include "u128.cuh"
+
+class u256 {
+public:
+  u256() = default;
+  __device__ constexpr u256(int l) : low(l), high(0) {}
+  __device__ constexpr u256(unsigned long l) : low(u128(l)), high(0) {}
+  __device__ constexpr u256(u128 l) : low(l), high(0) {}
+  __device__ constexpr u256(bool b) : low(b), high(0) {}
+  __device__ constexpr u256(u128 h, u128 l) : low(l), high(h) {}
+  __device__ constexpr u256(unsigned long hh, unsigned long hl,
+                            unsigned long lh, unsigned long ll)
+      : low(u128(lh, ll)), high(u128(hh, hl)) {}
+
+  __device__ constexpr u256 operator+(const u256 rhs) const {
+    return u256(high + rhs.high + ((low + rhs.low) < low), low + rhs.low);
+  }
+
+  __device__ constexpr u256 operator+=(const u256 rhs) {
+    *this = *this + rhs;
+    return *this;
+  }
+
+  __device__ constexpr inline u256 operator-(const u256 rhs) const {
+    return u256(high - rhs.high - ((low - rhs.low) > low), low - rhs.low);
+  }
+
+  __device__ constexpr u256 operator-=(const u256 rhs) {
+    *this = *this - rhs;
+    return *this;
+  }
+
+  __device__ constexpr bool operator==(const u256 rhs) const {
+    return high == rhs.high && low == rhs.low;
+  }
+
+  __device__ constexpr bool operator!=(const u256 rhs) const {
+    return !(*this == rhs);
+  }
+
+  __device__ constexpr bool operator<(const u256 rhs) const {
+    return ((high == rhs.high) && (low < rhs.low)) || (high < rhs.high);
+  }
+
+  __device__ constexpr u256 operator&(const u256 rhs) const {
+    return u256(high & rhs.high, low & rhs.low);
+  }
+
+  __device__ constexpr bool operator>(const u256 rhs) const {
+    return ((high == rhs.high) && (low > rhs.low)) || (high > rhs.high);
+  }
+
+  __device__ constexpr bool operator>=(const u256 rhs) const {
+    return !(*this < rhs);
+  }
+
+  __device__ constexpr bool operator<=(const u256 rhs) const {
+    return !(*this > rhs);
+  }
+
+  __device__ constexpr inline u256 operator>>(unsigned shift) const {
+    // TODO: reduce branch conditions
```

Comment:
That can be done as a sum of products with the bools I think.

Reply:
Something like:
```cpp
u128 new_low = (shift == 0) * low | (shift == 128) * high | (shift < 128 ^ shift == 0) * (high << (128 - shift) | (low >> shift)) | (256 > shift & shift > 128) * (high >> (shift - 128));
u128 new_high = (shift == 0) * high | (shift < 128 ^ shift == 0) * (high >> shift);
return u256(new_high, new_low);
```

Reply:
Thanks for the hack! I added it, but there are some parts of the code that could use a similar approach. As it's an optimization, I'm gonna leave that change for future PRs.